### ISSUE TITLE
errors when parsing docker-compose files with service-definitions with no image stanza (return of #97)

### ISSUE THREAD

**Issue by andoks** [safe-waters/docker-lock#111]
#97 seems to have returned in v0.8.8. I downgraded to v0.8.5, and the exact same project now works.

> Currently docker-lock generate errors out when parsing docker-compose files which contains service-definitions with no image stanza.
>
> I have "composition" compose-files, where we pass multiple compose files to docker-compose (e.g docker-compose --file base-system.docker-compose.yml --file host-specific-options.docker-compose.yml). The "host-specific-options.docker-compose.yml" file in this example may contain only the volumes and port bindings for all the service entries defined in "base-system.docker-compose.yml". Do note that the "addon" docker-compose.yml files may contain image stanzas for certain services (or additional services altogether).
>
> Is it possible to have docker-lock not error out on service-definitions that does not contain an image stanza (and perhaps log or error out if the image stanza is present, but not a valid image)?

**Comment by michaelperel** [comment]
TLDR: For now, this works in master. To test it out in the directory with your docker-compose file, use the docker image built from the CI/CD for master:

```docker run -v "${PWD}":/run safewaters/docker-lock lock generate```

The reason that this broke was that I switched to the new official parser in the last release, and the behavior was slightly different. I thought I had tests for it, but did not.

I fixed it in master by forking the official parser and making a very slight change that skips validation just like before (they already had the code to skip validation, it just wasn’t publicly exposed).

[I am working to get that change merged into the official parser](https://github.com/compose-spec/compose-go/issues/168), so that I don’t have to maintain a fork. The point of using the official parser is so that when new features get added, docker-lock will automatically support them.

Sorry for the inconvenience, and I will cut a new release once I get a response to my changes in the official parser repo.

**Comment by michaelperel** [comment]
In case the maintainers do not respond to me, I have built a similarly working version with their lower level APIs in the branch `miperel/without-fork`. I have no intention of maintaining my own fork, I would just like to use their highest level APIs to make sure `docker-lock` is least impacted by upstream changes.

Either way, a new release will be cut soon fixing the issue.

**Comment by andoks** [comment]
I tested v0.8.9 just now to test if this has been fixed, but I still get the error `'docker-compose.yml' failed to parse with err: service "influxdb" has neither an image nor a build context specified: invalid compose project`. Should I expect this to work now?

**Comment by michaelperel** [comment]
@andoks I tested it with the example in #97 and it worked, but upon further testing, the new official parser does choke on a few things that the old official parser does not, even when told to skip validation.

```
version: ‘3.0’
services:
  web:
```

is one such example.

I will look into what I can do to make the official parser not choke on something like this (I did have the previous PR merged upstream).

Thanks for reporting your experience. I figured since the example I had written in the previous issue works that it had been fixed, but it is always a bit hard to find edge cases when I didn’t write the parser. I do think this pain is worth it, to use the official parser.

**Comment by michaelperel** [comment]
[Here is where it fails](https://github.com/compose-spec/compose-go/blob/8ad38fd1e53f37e7486c139fe5b4579d3e531246/loader/loader.go#L214) if you are curious, so hopefully it should be as easy as skipping that check. I will know later today, and can hopefully post a fix shortly.

**Comment by michaelperel** [comment]
@andoks Could you post an example of what fails for you?

**Comment by michaelperel** [comment]
closed with a fix for a variety of files I tried, the next release which will be cut shortly should work.

**Comment by andoks** [comment]
@michaelperel Apologies for the late response. I will test with the newest version, and report back if it worked or not. And if not, try to produce a minimal repro.

**Comment by andoks** [comment]
I can confirm that v0.8.10 now works on the project files v0.8.9 failed for. Thank you very much!  :smile:
